src/bjm.h

```diff
@@ -0,0 +1,63 @@
+/*
+ * Background Job Manager - submit jobs to a background thread.
+ */
+
+#ifndef __BJM_H__
+#define __BJM_H__
+
+#include "sds.h"
+
+
+/* Initialize BJM with the requested number of background threads.
+ */
+void bjmInit(int numThreads);
+
+
+/* Provided job functions will be executed on a different thread and passed the provided privdata.
+ */
+typedef void (*bjmJobFunc)(void *privdata);
+
+/* After registering a function, the returned function handle can be used to submit jobs.
+ */
+typedef int bjmJobFuncHandle;
+
+
+/* Register a job function which can process background jobs.  A handle is returned for submitting
+ * jobs & gathering metrics.  This function is idempotent - submitting the same function again will
+ * return the same handle.  Handle values will be > 0, so this pattern can be used:
+ *
+ *    static bjmJobFuncHandle myHandle;
+ *    if (!myHandle) myHandle = bjmRegisterJobFunc(myFunc);
+ *    bjmSubmitJob(myHandle, ...);
+ *
+ * This co-locates a static variable at the point of job submission, and avoids repeated
+ * registration calls.
+ */
+bjmJobFuncHandle bjmRegisterJobFunc(bjmJobFunc func);
```

Comment:
the register-and-submit pattern everywhere seems a little annoying, comparing with it I still prefer the previous way of static pre-defined enum, or we can just submit the func and hash it to specific slot, without a pre-register.

Reply:
> the register-and-submit pattern everywhere seems a little annoying

@uvletter I agree with you.  I actually coded it as a hash on my first draft.  I changed to the registration mechanism in order to avoid the need for a hash lookup on every submission.  The function registration is simple for units containing an init function.  Without an init function, we can either create one, or do the static registration trick.  A hash lookup isn't much, but when you add it up for every lazy eviction (and other uses), it seems pretty wasteful since it can be easily avoided with a registration.

Note that I don't like the static enum - at least not as it's used today.  This is one of the reasons that I'm proposing the refactor.  We shouldn't have to change BIO in order to have a new job to submit.